lbrynet/reflector/server/server.py

```diff
@@ -63,7 +66,23 @@ def clean_up_failed_upload(self, err, blob):
 
     @defer.inlineCallbacks
     def _on_completed_blob(self, blob, response_key):
-        yield self.blob_manager.blob_completed(blob)
+        should_announce = False
+        if response_key == RECEIVED_SD_BLOB:
+            sd_info = yield BlobStreamDescriptorReader(blob).get_info()
+            yield save_sd_info(self.stream_info_manager, sd_info)
+            yield self.stream_info_manager.save_sd_blob_hash_to_stream(sd_info['stream_hash'],
+                                                                       blob.blob_hash)
+            should_announce = True
+
+        else:
+            stream_hash = yield self.stream_info_manager.get_stream_of_blob(blob.blob_hash)
```

Comment:
Why announce if the blob isn't an sd blob? If the sd blob hasn't completed, then stream hash would be none and the blob would have `should_announce` as false. If the sd blob had completed, then it and each subsequent stream blob received would have `should_announce` is true. If that's the desired behavior, then it should probably also go through any data blobs received before the sd blob completed and mark should_announce as true for them too.

Reply:
should_announce is set to True if its a SD blob or if its a head blob (blob_num == 0) 

Since the first line of the function sets should_announce to False, all other blobs should have should_announce == False 

I don't think its possible to receive blobs from clients before receiving the sd blobs when receiving streams.

Reply:
If the head blob was sent via reflector v1 before the sd blob, then only the sd blob would be announced. If the sd blob was also sent via reflector v1 it wouldn't be announced either, since the response key would be RECEIVED_BLOB.

These could be fixed by checking that the sd blob is known by the stream info manager in `handle_descriptor_request` and by scheduling an immediate_announce callLater for the head blob if it already exists when `_on_completed_blob` is called for a sd blob.